python/cuml/manifold/simpl_set.pyx

```diff
@@ -0,0 +1,367 @@
+        intersection.
+    local_connectivity: int (optional, default 1)
+        The local connectivity required -- i.e. the number of nearest
+        neighbors that should be assumed to be connected at a local level.
+        The higher this value the more connected the manifold becomes
+        locally. In practice this should be not more than the local intrinsic
+        dimension of the manifold.
+    verbose: bool (optional, default False)
+        Whether to report information on the current progress of the algorithm.
+    Returns
+    -------
+    fuzzy_simplicial_set: coo_matrix
+        A fuzzy simplicial set represented as a sparse matrix. The (i,
+        j) entry of the matrix represents the membership strength of the
+        1-simplex between the ith and jth sample points.
+    """
+
+    deterministic = random_state is not None
+    if not isinstance(random_state, int):
+        if isinstance(random_state, np.random.RandomState):
+            rs = random_state
+        else:
+            rs = np.random.RandomState(random_state)
+        random_state = rs.randint(low=np.iinfo(np.int32).min,
+                                  high=np.iinfo(np.int32).max,
+                                  dtype=np.int32)
+
+    cdef UMAPParams* umap_params = new UMAPParams()
+    umap_params.n_neighbors = <int> n_neighbors
+    umap_params.random_state = <uint64_t> random_state
+    umap_params.deterministic = <bool> deterministic
+    umap_params.set_op_mix_ratio = <float> set_op_mix_ratio
+    umap_params.local_connectivity = <float> local_connectivity
+    umap_params.verbosity = <int> verbose
+
+    X_m, n_rows, n_cols, dtype = \
+        input_to_cuml_array(X, order='C', check_dtype=np.float32)
+
+    handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+    cdef unique_ptr[COO] fss_graph = get_graph(handle_[0],
+                                               <float*><uintptr_t> X_m.ptr,
+                                               <float*><uintptr_t> NULL,
+                                               <int> X_m.shape[0],
+                                               <int> X_m.shape[1],
+                                               <UMAPParams*> umap_params)
+
+    cdef int nnz = fss_graph.get().nnz
+    coo_arrays = []
+    for ptr, dtype in [(<uintptr_t>fss_graph.get().vals(), np.float32),
+                       (<uintptr_t>fss_graph.get().rows(), np.int32),
+                       (<uintptr_t>fss_graph.get().cols(), np.int32)]:
+        mem = cp.cuda.UnownedMemory(ptr=ptr,
+                                    size=nnz * 4,
+                                    owner=None,
+                                    device_id=-1)
+        memptr = cp.cuda.memory.MemoryPointer(mem, 0)
+        coo_arrays.append(cp.ndarray(nnz, dtype=dtype, memptr=memptr))
+
+    result = cp.sparse.coo_matrix(((coo_arrays[0],
+                                    (coo_arrays[1], coo_arrays[2]))))
+    result = result.copy()
```

Comment:
It's unfortunate that we need to copy here. Can you add a comment/todo in the code here that we should revisit this in the future? We're going to be redesigning our sparse objects in RAFT and I'd like to see if we can remove the need for the copy here (e.g. using heap-based allocation and allowing cython to take ownership).